lib/matplotlib/testing/decorators.py

```diff
@@ -227,43 +236,66 @@ def _pytest_image_comparison(baseline_images, extensions, tol,
     Decorate function with image comparison for pytest.
 
     This function creates a decorator that wraps a figure-generating function
-    with image comparison code. Pytest can become confused if we change the
-    signature of the function, so we indirectly pass anything we need via the
-    `mpl_image_comparison_parameters` fixture and extra markers.
+    with image comparison code.
     """
     import pytest
 
     extensions = map(_mark_skip_if_format_is_uncomparable, extensions)
+    KEYWORD_ONLY = inspect.Parameter.KEYWORD_ONLY
 
     def decorator(func):
+        old_sig = inspect.signature(func)
+
         @functools.wraps(func)
-        # Parameter indirection; see docstring above and comment below.
-        @pytest.mark.usefixtures('mpl_image_comparison_parameters')
         @pytest.mark.parametrize('extension', extensions)
-        @pytest.mark.baseline_images(baseline_images)
-        # END Parameter indirection.
         @pytest.mark.style(style)
         @_checked_on_freetype_version(freetype_version)
         @functools.wraps(func)
-        def wrapper(*args, **kwargs):
+        def wrapper(*args, extension, request, **kwargs):
             __tracebackhide__ = True
+            if 'extension' in old_sig.parameters:
+                kwargs['extension'] = extension
+            if 'request' in old_sig.parameters:
+                kwargs['request'] = request
+
             img = _ImageComparisonBase(func, tol=tol, remove_text=remove_text,
                                        savefig_kwargs=savefig_kwargs)
             matplotlib.testing.set_font_settings_for_testing()
             func(*args, **kwargs)
 
-            # Parameter indirection:
-            # This is hacked on via the mpl_image_comparison_parameters fixture
-            # so that we don't need to modify the function's real signature for
-            # any parametrization. Modifying the signature is very very tricky
-            # and likely to confuse pytest.
-            baseline_images, extension = func.parameters
-
-            assert len(plt.get_fignums()) == len(baseline_images), (
+            # If the test is parametrized in any way other than applied via
+            # this decorator, then we need to use a lock to prevent two
+            # processes from touching the same output file.
+            needs_lock = any(
+                marker.args[0] not in {'extension'}
+                for marker in request.node.iter_markers('parametrize'))
+
+            if baseline_images is not None:
```

Comment:
just do `if baseline_images is None: baseline_images = request.getfixturevalue(...)`?

Reply:
That causes a read-before-use error on the variable, though maybe it'd work if marked non-local? It didn't need to modify the external variable so I didn't try that.

Reply:
Ah, sure, no worries.